docs/index.bs

```diff
@@ -3068,88 +3049,74 @@ spec: webappsec-referrer-policy; urlPrefix: https://w3c.github.io/webappsec-refe
+          1. [=map/For each=] |worker| → |state| of |workerUpdates|:
+              1. Let |workerObject| be the {{ServiceWorker}} that represents |worker| within |settingsObject|'s [=environment settings object/realm execution context=].
+              1. If |workerObject|'s {{ServiceWorker/state}} is not |state|, then:
+                  1. Set |workerObject|'s {{ServiceWorker/state}} to |state|.
+                  1. If |state| is not *installing*, then [=list/append=] |workerObject| to |changedWorkers|.
+
+                      Note: *installing* is the initial script-exposed state, so it doesn't count as a change.
+
+          1. [=list/For each=] |registrationObject| of |registrationObjects|:
+              1. Set |registrationObject|'s {{ServiceWorkerRegistration/installing}} to null if |installingWorker| is null, otherwise set it to the {{ServiceWorker}} that represents |installingWorker| within |settingsObject|'s [=environment settings object/realm execution context=].
+              1. Set |registrationObject|'s {{ServiceWorkerRegistration/waiting}} to null if |waitingWorker| is null, otherwise set it to the {{ServiceWorker}} that represents |waitingWorker| within |settingsObject|'s [=environment settings object/realm execution context=].
+              1. Set |registrationObject|'s {{ServiceWorkerRegistration/active}} to null if |activeWorker| is null, otherwise set it to the {{ServiceWorker}} that represents |activeWorker| within |settingsObject|'s [=environment settings object/realm execution context=].
+          1. [=list/For each=] |workerObject| of |changedWorkers|, [=fire an event=] on |workerObject| named `statechange`.
+          1. If `updateFound` is true, then [=list/for each=] |registrationObject| of |registrationObjects|, [=fire an event=] on |registrationObject| named `updatefound`.
+          1. If |newActiveWorker| is true:
+              1. Let |container| be the {{ServiceWorkerContainer}} object within |settingsObject|'s [=environment settings object/realm execution context=].
+              1. Let |readyPromise| be |container|'s [=ServiceWorkerContainer/ready promise=].
+              1. If |settingsObject| [=/uses=] |registration|, then:
+                  1. Set |settingsObject|'s [=environment/active service worker=] to |activeWorker|.
+                  1. [=fire an event=] named `controllerchange` on |container|.
+              1. <span id="update-state-resolve-ready-step">If the [=Match Service Worker Registration|matching=] [=/service worker registration=] for |settingsObject|'s [=environment/creation URL=] is |registration|, and |readyPromise| is pending, then [=/resolve=] |readyPromise| with a new {{ServiceWorkerRegistration}} object representing |registration| in |readyPromise|'s [=relevant Realm=].</span>
```

Comment:
This shouldn't be a new {{ServiceWorkerRegistration}}. It should return the same object if one exists. This is kinda hand-wavey in the spec.